### ISSUE TITLE
[Feature] Implicit Default Value On Optional Parameter

### ISSUE THREAD

**Issue by AdamSpeight2008** [AdamSpeight2008/roslyn-AdamSpeight2008#9]
# Implicit Default Value On Optional Parameter

The aim of the proposal is to see if it possible to make the **default value**, that currently as to explicitly expressed can be made optional.
### Examples

The existing code

``` vb.net
Public Function Example( Optional x As Integer = 0, Optional s As String = Nothing )

End Function
```

can written as

``` vb.net
Public Function Example( Optional x As Integer, Optional s As String )

End Function
```

This would reduce a lot of boiler-plate code for the likely common usage of just needing the default value of the type (eg `parameter = default(T)` ).
Will still permitting the expressing of a different value, should the coder require one.

---
# Research
## Grammar Definition

```
Parameter
    : Attributes? ParameterModifier* ParameterIdentifier ( 'As' TypeName )?
      ( Equals ConstantExpression )?
    ;
```

So the Visual Basic Grammer (VB14) allows for an optional default value `( Equals ConstantExpression )?`.
### Errors

`Default Value For Non Optional Parameter`

``` vb
Method( x As Integer = 0)
```

`Obsolete Optional Without Value`

``` vb
Method( Optional x As Integer)
```

==Compat== Must maintain this when language version doesn't support Implicit Default Value On Optional Parameter.

**Caller Information Attributes**

``` vb
( <CallerLineNumber> Optional line As Integer = 0 )
( <CallerMethodName> Optional method As String = Nothing )
( <CallerFilename>   Optional filename As String = Nothing )
```

**(Compat)** These should continue work, even when using the feature.

---
## **Parser**

After a search of the Parser code
- [ParseParameter](http://source.roslyn.io/#Microsoft.CodeAnalysis.VisualBasic/Parser/Parser.vb,4576)

``` vb
Dim equals As PunctuationSyntax = Nothing
Dim value As ExpressionSyntax = Nothing

' TODO - Move these errors (ERRID.ERR_DefaultValueForNonOptionalParamout, ERRID.ERR_ObsoleteOptionalWithoutValue) of the parser. 
' These are semantic errors. The grammar allows the syntax. 
If TryGetTokenAndEatNewLine(SyntaxKind.EqualsToken, equals) Then

    If Not (modifiers.Any AndAlso modifiers.Any(SyntaxKind.OptionalKeyword)) Then
        equals = ReportSyntaxError(equals, ERRID.ERR_DefaultValueForNonOptionalParam)
    End If

    value = ParseExpressionCore()

ElseIf modifiers.Any AndAlso modifiers.Any(SyntaxKind.OptionalKeyword) Then

    equals = ReportSyntaxError(InternalSyntaxFactory.MissingPunctuation(SyntaxKind.EqualsToken), ERRID.ERR_ObsoleteOptionalWithoutValue)
    value = ParseExpressionCore()

End If
```

Make the body of the second condition dependant on the feature's availability.

``` vb
Dim initializer As EqualsValueSyntax = Nothing

If value IsNot Nothing Then

    If value.ContainsDiagnostics Then
        value = ResyncAt(value, SyntaxKind.CommaToken, SyntaxKind.CloseParenToken)
    End If

    initializer = SyntaxFactory.EqualsValue(equals, value)
End If

Return SyntaxFactory.Parameter(attributes, modifiers, paramName, optionalAsClause, initializer)

```

Depending on the code change, the following could be null `value = nothing` and `equals = nothing`.
We would require to modify these conditions. 
- [ParseParameters](http://source.roslyn.io/#Microsoft.CodeAnalysis.VisualBasic/Parser/Parser.vb,4419)
  
  I do not see any potential issue within this function, when using the feature.
- [ParseParameterList](http://source.roslyn.io/#Microsoft.CodeAnalysis.VisualBasic/Parser/Parser.vb,3671)

I do not see any potential issue within this function, when using the feature.
## **Syntax**

**`SyntaxFactory.Parameter`**
### Overload Resolution

[OverloadResolution.OptionalArgument](http://source.roslyn.io/#Microsoft.CodeAnalysis.VisualBasic/Semantics/OverloadResolution.vb,509) **(Potential Issue)**
- What if `DefaultValue = Nothing` ?
- Is it expected to be non-null?
  
  Handle [Optional Parameters](http://source.roslyn.io/#Microsoft.CodeAnalysis.VisualBasic/Semantics/OverloadResolution.vb,3056) **(Potential Issue)**
- If this returns `null` then should I do check for the feature and assign "default" value to `default argument`?
## **Symbol**

[ParameterSymbol](http://source.roslyn.io/#Microsoft.CodeAnalysis.VisualBasic/Symbols/ParameterSymbol.vb,138) **(Potential Issue)**

``` vb
''' <summary>
''' Returns the default value of this parameter. If <see cref="HasExplicitDefaultValue"/>
''' returns false, then this property throws an InvalidOperationException.
''' </summary>
Public ReadOnly Property ExplicitDefaultValue As Object
    Get
        If HasExplicitDefaultValue Then
            Return ExplicitDefaultConstantValue.Value
        End If

        Throw New InvalidOperationException
    End Get
End Property
```

This feature would not have an explicit default value, as it let's the coder not declare one.
So the options are:-
- Add an additional condition based on the on feature's availiblity, to return a different value.
- At the usage sites, catch the exception then, modify the usage.
  
  [Emit.ParameterSymbolAdapter](http://source.roslyn.io/#Microsoft.CodeAnalysis.VisualBasic/Emit/ParameterSymbolAdapter.vb,f2e85a004c64ee62) **(Potential Issue)**
  - `HasMetadateConstantValue`
  - `GetMetadataConstantValue`
  - `IParameterDefinition_HasDefaultValue`
  
  [Emit.NoPia.EmbeddedParameter](http://source.roslyn.io/#Microsoft.CodeAnalysis.VisualBasic/Emit/NoPia/EmbeddedParameter.vb) **(Potential Issue)**
  - `HasDefaultValue`
  - `GetDefaultValue`
## **Binding**

[Binding.Binder_Invocation](http://source.roslyn.io/#Microsoft.CodeAnalysis.VisualBasic/Binding/Binder_Invocation.vb,3025) **(Potential Issue)**
- Comment `HasDefaultValue` indicates `(Parameter.IsOptional == True) && (Parameter.DefaultValue <> null)`
- Deals with **Caller Info Attributes** eg `<CallerLineNumber>`, `<CallerMethodName>`, `<CallerFilename>`
- `ElseIf param.IsOptional Then` [Here](http://source.roslyn.io/#Microsoft.CodeAnalysis.VisualBasic/Binding/Binder_Invocation.vb,3153)
  - Suggests that this branch is taken on `( Optional p As Integer )`
  - If it is the **Caller Info Attributes** are ignored (_bug?_)
- [BindParameterDefaultValue](https://github.com/dotnet/roslyn/blob/586f37ae8ee27162ea334a22d6b431cdd5e05a14/src/Compilers/VisualBasic/Portable/Binding/Binder_Initializers.vb#L641)
- [InitializerSemanticModule.BindInitializer](http://source.roslyn.io/#Microsoft.CodeAnalysis.VisualBasic/Binding/InitializerSemanticModel.vb,139)

**Comment by ljw1004** [comment]
I really like this feature. @AnthonyDGreen what do you think? @AdamSpeight2008 There are a few further areas of research to investigate...
1. Metadata export. How would we export such a thing?
2. Metadata import. I seem to remember that the VB compiler currently imports optional parameters that lack values, with the semantics you describe. Is this right? (Why? Under what circumstances? What exactly is the encoding in metadata?)
3. COM. I seem to remember that COM is the chief area where there might be metadata that has optional parameters without default values. Is this right?

**Comment by AdamSpeight2008** [comment]
@ljw1004 
The current `emit` and `metadata` unit test pass.
The major block is [#12690](https://github.com/dotnet/roslyn/issues/12690), which is affects the `ExpressionEvaluator` unit tests.
Overload resolution issue I think I fixed in [#12629](. commit: [link](https://github.com/dotnet/roslyn/pull/12629/commits/1039dfb237ee73b9c69438f67cc5539f65993027)

Analyser and CodeFix need a way of validating the presence of the Feature flag. (currently not of github)
Or I could do a sample parse and check for the error, `BC30812 Optional parameters must specify a default value`. _I think that is kind of hacky_

**Comment by AdamSpeight2008** [comment]
@ljw1004 @AnthonyDGreen

Reran the unit test after merging in [Roslyn PR](https://github.com/dotnet/roslyn/pull/12772).
They pass (No Failures) locally, waiting on the results of the Roslyn PR Build.
When they pass I'll do a tidy up of the code to reduce the size of the pull request.
Refactor some of the method names.
